### PR TITLE
Test serialization for serial and shared triangulations

### PR DESCRIPTION
**Pull request by pcafrica**
Depends on #15576.

Tests for serializing/deserializing were already implemented for `p::d::T` and `p::f::T`.

The new design introduced in #15576 unlocks the same interface also for `T` and `p::s::T`.

This PR adds 3 new tests for serializing/deserializing:
- a serial triangulation
- a serial triangulation and an associated FE vector
- a shared triangulation and an associated FE vector, on 2 MPI ranks

@peterrum @blaisb @bangerth

### CODE REVIEW

**Comment by peterrum** [comment]
@pcafrica After you rebase, I'll take a look at this PR.

**Comment by pcafrica** [comment]
Thank you @peterrum! I have just rebased, this MR is ready for review.

File: tests/fullydistributed_grids/copy_distributed_tria_01.cc [hunk 1]
```diff
@@ -30,7 +30,7 @@
 #include <deal.II/grid/tria.h>
 #include <deal.II/grid/tria_description.h>
 
-#include "./tests.h"
+#include "../grid/tests.h"
```
Comment (peterrum) [inline tests/fullydistributed_grids/copy_distributed_tria_01.cc:33]: ?

File: tests/grid/tests.h [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+#ifndef dealii_grid_tests_h
+#define dealii_grid_tests_h
+
+#include "../tests.h"
+
+
+template <int dim, int spacedim>
+void
+print_statistics(const Triangulation<dim, spacedim> &tria, bool do_mg = false)
```
Comment (peterrum) [inline tests/grid/tests.h:24]: It seems that you have not removed these function from the `fullydistributed_grid` folder!?

File: tests/grid/serialization_01.cc [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+  VectorType vector_loaded(dof_handler.n_dofs());
+
+  print_statistics(triangulation, false);
+
+  std::stringstream stream;
+  {
+    boost::archive::text_oarchive oa(stream);
+    oa << triangulation;
+    oa << vector;
```
Comment (peterrum) [inline tests/grid/serialization_01.cc:74]: I don't think we need to test here the vector. There should be other tests whether the serialization/deserialization of vectors work.

File: tests/grid/save_load_01.cc [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+using namespace dealii;
+
+template <int dim, typename TriangulationType>
+void
+test(TriangulationType &triangulation)
+{
+  const std::string filename = "save_load_" + std::to_string(dim) + "d_out";
+
+  triangulation.save(filename);
```
Comment (peterrum) [inline tests/grid/save_load_01.cc:43]: But here we should test vectors sometime in the future ;) What is missing for that?

File: tests/sharedtria/serialization_01.cc [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  VectorTools::interpolate(dof_handler, InterpolationFunction<dim>(), vector);
+
+  VectorType vector_loaded(dof_handler.n_dofs());
+
+  print_statistics(triangulation, false);
+
+  std::stringstream stream;
+  {
+    boost::archive::text_oarchive oa(stream);
+    oa << triangulation;
```
Comment (peterrum) [inline tests/sharedtria/serialization_01.cc:73]: I am wondering what happens for `p:s:T` for load/save? I guess it should work but users need to provide unique file names on each process!?

**Comment by peterrum** [comment]
@pcafrica  One thing I am wondering of is whether `parallel:distributed::SolutionTransfer` now also works serial triangulation (at least the serialization/deserialization part).

File: tests/fullydistributed_grids/copy_distributed_tria_01.cc (hunk 1 above)
Comment (pcafrica) [reply tests/fullydistributed_grids/copy_distributed_tria_01.cc:33]: The `tests.h` file has just been moved from `fullydistributed_grids` to `grids`, without modify its content.

Is it fine to reference it also from other folders other than `grids`?

File: tests/grid/save_load_01.cc (hunk 4 above)
Comment (pcafrica) [reply tests/grid/save_load_01.cc:43]: Nothing is missing, indeed :)

This is actually tested in `grids/serialization_01.cc`.

**Comment by peterrum** [comment]
```
/jenkins/workspace/dealii-serial_PR-15649/tests/serialization/parallel_fullydistributed_construction_data_1.cc:29:10: fatal error: ../fullydistributed_grids/tests.h: No such file or directory
   29 | #include "../fullydistributed_grids/tests.h"
```

**Comment by pcafrica** [comment]
> ```
> /jenkins/workspace/dealii-serial_PR-15649/tests/serialization/parallel_fullydistributed_construction_data_1.cc:29:10: fatal error: ../fullydistributed_grids/tests.h: No such file or directory
>    29 | #include "../fullydistributed_grids/tests.h"
> ```

Fixed, thank you!

**Comment by tamiko** [comment]
@peterrum Do you approve now?

File: tests/sharedtria/serialization_01.cc (hunk 5 above)
Comment (peterrum) [reply tests/sharedtria/serialization_01.cc:73]: Regarding step-83: I think `triangulation.save(filename);` is the right approach.

**Review by peterrum — APPROVED** [review]

**Pull request merged**